# Mini-Project - Validation

##### Student Tags

Author: Anderson Hitoshi Uyekita    
Mini-Project: Validation  
Course: Data Science - Foundations II  
COD: ND111  
Date: 24/01/2019    

***

## Table of Contents
- [Introduction](#intro)
- [Given code 1](#code1)
- [Exercise 1](#part_i_1)
- [Exercise 2](#part_i_2)

***

## General Information

This Jupyter Notebook (in Python 2) aims to create a reproducible archive.

## Introduction <a id='intro'></a>

In this mini-project, you’ll start from scratch in making a training-testing split in the data. This will be the first step toward your final project, of building a POI identifier.

## Exercise 1 - Your First (Overfit) POI Identifier <a id='part_i_1'></a>

You’ll start by building the simplest imaginable (unvalidated) POI identifier. The starter code (validation/validate_poi.py) for this lesson is pretty bare--all it does is read in the data, and format it into lists of labels and features. Create a decision tree classifier (just use the default parameters), train it on all the data (you will fix this in the next part!), and print out the accuracy. THIS IS AN OVERFIT TREE, DO NOT TRUST THIS NUMBER! Nonetheless, what’s the accuracy?

<em>From Python 3.3 forward, a change to the order in which dictionary keys are processed was made such that the orders are randomized each time the code is run. This will cause some compatibility problems with the graders and project code, which were run under Python 2.7. To correct for this, add the following argument to the featureFormat call on line 25 of validate_poi.py:

sort_keys = '../tools/python2_lesson13_keys.pkl'

This will open up a file in the tools folder with the Python 2 key order.

Note: If you are not getting the results expected by the grader, then you may want to check the file tools/feature_format.py. Due to changes in the final project, some file changes have affected the numbers output on this assignment as written. Check that you have the most recent version of the file from the repository, such that the featureFormat has a default parameter for sort_keys = False and that keys = dictionary.keys() results.</em>

In [1]:
#!/usr/bin/python


"""
    Starter code for the validation mini-project.
    The first step toward building your POI identifier!

    Start by loading/formatting the data

    After that, it's not our code anymore--it's yours!
"""

import pickle
import sys
sys.path.append("../tools/")
from feature_format import featureFormat, targetFeatureSplit

data_dict = pickle.load(open("../final_project/final_project_dataset.pkl", "r") )

### first element is our labels, any added elements are predictor
### features. Keep this the same for the mini-project, but you'll
### have a different feature list when you do the final project.
features_list = ["poi", "salary"]

data = featureFormat(data_dict, features_list)
labels, features = targetFeatureSplit(data)

### it's all yours from here forward!  

# Importing the Scikit Learn Decision Tree Module.
from sklearn import tree

# Creating the Classifier.
clf = tree.DecisionTreeClassifier()

# Fitting/Training with ALL observations.
clf = clf.fit(features, labels)

# Predicting using the same dataset. OVERFITTING!!
pred = clf.predict(features)

# Importing numpy.
import numpy as np

# Importing the Accuracy module.
from sklearn.metrics import accuracy_score

# Calculating the accuracy of the overfitted model.
acc = accuracy_score(labels, pred)

In [2]:
print "Accuracy:", acc

Accuracy: 0.9894736842105263


>**What is the accuracy of your first (overfit) POI identifier?**

98.95%

## Exercise 2 - Deploying a Training/Testing Regime <a id='part_i_2'></a>

Now you’ll add in training and testing, so that you get a trustworthy accuracy number. Use the train_test_split validation available in sklearn.cross_validation; hold out 30% of the data for testing and set the random_state parameter to 42 (random_state controls which points go into the training set and which are used for testing; setting it to 42 means we know exactly which events are in which set, and can check the results you get). What’s your updated accuracy?

In [3]:
# Importing the model selection module.
from sklearn.model_selection import train_test_split

# Splitting the dataset into train and test.
features_train, features_test, labels_train, labels_test = train_test_split(features, labels,test_size=0.30, random_state=42)

### it's all yours from here forward!  

# Importing the Scikit Learn Decision Tree Module.
from sklearn import tree

# Creating the Classifier.
clf = tree.DecisionTreeClassifier()

# Fitting/Training with ALL observations.
clf = clf.fit(features_train, labels_train)

# Predicting using the same dataset. OVERFITTING!!
pred = clf.predict(features_test)

# Importing numpy.
import numpy as np

# Importing the Accuracy module.
from sklearn.metrics import accuracy_score

# Calculating the accuracy of the overfitted model.
acc = accuracy_score(labels_test, pred)

In [4]:
# Printing the accuracy
acc

0.7241379310344828

>**What is the accuracy with a testing regime properly deployed?**

72.41%

#### Copying file

In [5]:
# Importing shutil to deal with copy
from shutil import copyfile

# File name
filename = 'validate_poi.ipynb'

# Lesson
lesson = '14-Lesson_14'

# Directory to make a copy
dir_copy = '../../' + lesson + '/00-Mini Project/' + filename

# Copying file.
copyfile(filename, dir_copy)